### ISSUE TITLE
Performance regression in GTR2 VR on Windows possibly caused by changes between 2/18 and 2/19

### ISSUE THREAD

**Issue by TheIronWolfModding** [doitsujin/dxvk#2505]
I have been testing changes done on master between 2/18 and 2/19 (lockless stuff) merged into my fork, to my surprise I am seeing regression in my scenario where I am using DXVK in DX9 mode in VR.  In heavy CPU intense scene framerate dropped from 58FPS to 47FPS.  Before I collect more data I need to know if it is needed, because I am on Win10 and very convoluted setup.

**Reports with no log files will be ignored.**

### Software information
Name of the game, settings used etc.

### System information
- GPU:
- Driver:
- Wine version: 
- DXVK version: 

### Apitrace file(s)
- Put a link here

### Log files
- d3d9.log:
- d3d11.log:
- dxgi.log:

**Comment by K0bin** [comment]
Please make an apitrace so we can actually investigate this. We also need the log files you left out.

I have no idea how you even use DXVK D3D9 in VR and I don't own the game, so we also need you to [bisect](https://git-scm.com/docs/git-bisect) the performance regression down to a commit.

**Comment by TheIronWolfModding** [comment]
Thanks for your response.  That's the problem, I use slightly modified DXVK that I build myself, it's not exactly the same as main branch.     That's why I did not provide logs (yet)

I want to help, I am just not sure given that I am running Windows and my own fork is it even worth collecting stuff?  Will apitrace still help or am I on my own here?

**Comment by doitsujin** [comment]
I somehow strongly doubt the lockless stuff itself can be responsible here since it's just traversing a list that's always uncontested during actual gameplay anyway. Only thing I could imagine is that the game compiles hundreds of variants of certain pipelines and doesn't like that the order of those is reversed now.

**Comment by TheIronWolfModding** [comment]
I agree with you, there have been a lot of changes and it's possible I made a mistake on rebuilding/merging.  

Yesterday I realized that I need to reproduce regression in 2D, based on main branch binary and if I can still detect it I'll get apitrace, I'll try that next week.  Thanks for your time.

**Comment by TheIronWolfModding** [comment]
I got a chance to capture a trace.   I was also able to confirm regression repros in with code without my changes at all, master, and is caused by commit(s) between 8e116303704a4e07179 and 80f744549f1c8. 

Trace: https://1drv.ms/u/s!ApNCoFDyNnCJhMtq1EBazaktQe1nRA?e=4nKZCQ 
Log: https://1drv.ms/u/s!ApNCoFDyNnCJhMtrMXD42-yDk1JEHw?e=zibKB2

I also think that with latest changes message  "Compiling shaders" stays in HUD longer, but I need to test much more...

**Comment by doitsujin** [comment]
I can't open the archive, what software was used to create it?

**Comment by TheIronWolfModding** [comment]
Looks like I uploaded incomplete archive, sorry, re-uploading (it's .zip created by 7z)

**Comment by TheIronWolfModding** [comment]
Here it is: https://1drv.ms/u/s!ApNCoFDyNnCJhMtq1EBazaktQe1nRA?e=4IGeSf

**Comment by doitsujin** [comment]
Thanks, this one works. And yeah, apitrace is its own thing.

**Comment by TheIronWolfModding** [comment]
this effort is a mess - game is old and outdated, it has hard time starting in regular DX9 mode.  I have plugin that fixes several startup issues, and part of that plugin is in d3d9.dll proxy.  So, apitrace is not exactly how game is used, but hopefully it helps.  Please, let me know how I can help further (although I may not have time till Thursday).   And, thank you very much for DXVK, it is amazing and will allow to keep GTR2 alive for years to come.

**Comment by doitsujin** [comment]
I'm seeing up to 27 variants being created for a single pipeline, so yeah, there might be extra overhead when the game wants specific variants more often than others. Can you verify whether it really is 67e2ee1b2635f6867d3c92a14192bffb1c026d33 that's causing the issue?

**Comment by TheIronWolfModding** [comment]
Thanks for looking.  Yes, i will revert that commit when i get a chance and get back to you.

Btw this same graphics engine is used in dozen of racing simulators.  Based on what you seen in the trace, do you have any suggestions on how can I tweak dxvk to speed up that family of games?

**Comment by doitsujin** [comment]
I'm still trying to work out if the thing that I *think* might be causing a perf loss actually does so.

**Comment by doitsujin** [comment]
Ok, this can't really be it. The average number of lookups is **higher** with the previous code, at least during the actual gameplay segment. Only reason why it could still potentially be slower is CPU caching effects, but that's... 

Also worth noting that I haven't seen any performance regressions in D3D11 titles yet.

**Comment by TheIronWolfModding** [comment]
Thanks for looking.  The scene I am using to test is very much CPU bound.   Luckily I have only handful of commits to try, I'll let you know if I find anything.

**Comment by TheIronWolfModding** [comment]
I can confirm that performance drops ~20% as soon as I cherry pick 67e2ee1b2635f

My processor is 10900K with HT on.

**Comment by doitsujin** [comment]
Does https://github.com/doitsujin/dxvk/tree/lockless-test help?

**Comment by TheIronWolfModding** [comment]
Yes, it fixes the drop! (is it faster because loop is gone in insert*)?
